core/src/main/java/org/apache/shiro/subject/support/DelegatingSubject.java

```diff
@@ -471,7 +471,16 @@ public PrincipalCollection releaseRunAs() {
     private List<PrincipalCollection> getRunAsPrincipalsStack() {
         Session session = getSession(false);
         if (session != null) {
-            return (List<PrincipalCollection>) session.getAttribute(RUN_AS_PRINCIPALS_SESSION_KEY);
+            try {
+                return (List<PrincipalCollection>) session.getAttribute(RUN_AS_PRINCIPALS_SESSION_KEY);
+            } catch (SessionException se) {
+                // There could be a rare race condition when a session is invalidated in another thread,
+                // this thread could throw this exception, so we catch it
+                // similar issue as in clearRunAsIdentitiesInternal()
+                // See https://issues.apache.org/jira/browse/SHIRO-512
+                log.debug("Encountered session exception trying to get 'runAs' principal stack.  This "
```

Comment:
I'm not trying to be rude, I'm legitimately trying to understand, how is this safe to ignore? This appears to be the reason one of my unit tests is failing now after bumping Shiro up to `1.10.0`

So if Thread_1 invalidates the subject, then Thread_2 attempts to get the session, this block here would originally throw an exception to Thread_2 informing that the session has been invalidated. But now Thread_2 will catch and just log the exception instead. This would essentially let Thread_2 believe the subject is still authenticated, when the session should actually be invalidated.

Reply:
@adamenveil 
I can sort of see both sides of this.  I think what this change was trying to prevent was:

1. Thread 1 - calls `Session session = getSession(false);` // valid session returned
1. Thread 2 - calls `session.invalidate()`; // somewhere else in the code
1. Thread 1 - calls `session.getAttribute(RUN_AS_PRINCIPALS_SESSION_KEY);` // that throws an exception, the session is invalid.
1. Thread 1 - the `getRunAsPrincipalsStack` method returns null

If this same flow of events happened in a different order:
1. Thread 1 - calls `Session session = getSession(false);` // valid session returned
1. Thread 1 - calls `session.getAttribute(RUN_AS_PRINCIPALS_SESSION_KEY);` // object returned
1. Thread 2 - calls `session.invalidate()`; // somewhere else in the code

The caller of `getRunAsPrincipalsStack()` doesn't know the session has been invalidated.

Or in the order of:

1. Thread 2 - calls `session.invalidate()`; // somewhere else in the code
1. Thread 1 - calls `Session session = getSession(false);` // null is returned
1. Thread 1 - the `getRunAsPrincipalsStack` method returns null


Can you give us an idea of what your code is doing maybe there is something else we could do?

Reply:
i.e. it's not specified that the subject is automatically logged out when session is invalidated.
Proper way (according to the API) to log a subject out is `SecurityUtils.getSubject().invalidate()` and not `session.invalidate()`

It is true that Shiro *could* add HttpSession listeners and automatically log the subject out if a session gets invalidated, but neither the old nor new code does this, perhaps this could be an additional feature in the future, but certainly not a showstopper